x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/RetentionPolicyToDeleteByQueryRequestConverter.java

```diff
@@ -0,0 +1,97 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.transform.transforms;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.reindex.AbstractBulkByScrollRequest;
+import org.elasticsearch.index.reindex.DeleteByQueryRequest;
+import org.elasticsearch.xpack.core.transform.transforms.DestConfig;
+import org.elasticsearch.xpack.core.transform.transforms.RetentionPolicyConfig;
+import org.elasticsearch.xpack.core.transform.transforms.SettingsConfig;
+import org.elasticsearch.xpack.core.transform.transforms.TimeRetentionPolicyConfig;
+import org.elasticsearch.xpack.core.transform.transforms.TransformCheckpoint;
+
+import java.time.Instant;
+
+/**
+ * Implementation of `retention_policy` configuration parameter.
+ *
+ * All implementations of `rentention_policy` are converted to a {@link DeleteByQueryRequest}, which is than executed by the indexer.
+ */
+public final class RetentionPolicyToDeleteByQueryRequestConverter {
+
+    public static class RetentionPolicyException extends ElasticsearchException {
+        RetentionPolicyException(String msg, Object... args) {
+            super(msg, args);
+        }
+    }
+
+    private RetentionPolicyToDeleteByQueryRequestConverter() {}
+
+    /**
+     * Build a {@link DeleteByQueryRequest} from a retention policy. The DBQ should run _after_ a new checkpoint has finished.
+     * The given checkpoint should be the one that just finished indexing, however the DBQ executes before this checkpoint
+     * gets exposed.
+     *
+     * @param retentionPolicyConfig The retention policy configuration
+     * @param settingsConfig settings to set certain parameters
+     * @param destConfig the destination config
+     * @param checkpoint The checkpoint that just finished
+     *
+     * @return a delete by query request according to the given configurations or null if no delete by query should be executed
+     */
+    static DeleteByQueryRequest buildDeleteByQueryRequest(
+        RetentionPolicyConfig retentionPolicyConfig,
+        SettingsConfig settingsConfig,
+        DestConfig destConfig,
+        TransformCheckpoint checkpoint
+    ) {
+        if (checkpoint == null || checkpoint.isEmpty()) {
+            return null;
+        }
+
+        DeleteByQueryRequest request = new DeleteByQueryRequest();
+
+        if (retentionPolicyConfig instanceof TimeRetentionPolicyConfig) {
+            request.setQuery(getDeleteQueryFromTimeBasedRetentionPolicy((TimeRetentionPolicyConfig) retentionPolicyConfig, checkpoint));
+        } else {
+            throw new RetentionPolicyException("unsupported retention policy of type [{}]", retentionPolicyConfig.getWriteableName());
```

Comment:
Wouldn't this cause problems when other types of policies are added in subsequent versions?

Reply:
Yeah it would? I think thats the idea. If you add a new policy, you sure better make sure to write a request converter. 

Throwing here is a good pattern. It enforces case handling.

Reply:
> Or RetentionPolicyConfig know everything about itself it could have a toDeleteByQueryRequest method

Intentionally I avoided that. We had (probably still have) a lot of logic in the config classes in the past and I had a hard time to separate config and logic.

I agree it looks a little dumb at the moment, its just 1 config with 1 implementation.